### PR TITLE
Storage query tool/runtime artifact retrieval

### PR DESCRIPTION
**Pull request by jcurtisk**
## What does this PR do?

This PR adds support for zipped runtime artifact retrieval within the Storage Query Tool.

## How was this PR tested?

Manual verification with extension of test suite coming.

## Example commands

User can search for all objects in a bucket:
tiaf_tools.py --s3-bucket "yourbucket"
This will display all objects in a bucket

User can search for all objects in a root folder of a bucket:
tiaf_tools.py --s3-bucket "yourbucket" --search-in "your_root"
This displays all objects in that directory

User can search for a specific historic_data.json.zip file by specifying directory, branch, build and test suite:
tiaf_tools.py --s3-bucket "yourbucket" --full-address "RootDir/Branch/Build/TestSuite/historic_data.json.zip"
This will check whether the file exists.

Tool also provides Create, Read, Update and Delete functionality. Examples use the json file type, but it is possible to use "zip" as well.

Create:
tiaf_tools.py --s3-bucket "tiaf" --action "create" --full-address "RootDir/Branch/Build/TestSuite/historic_data.json.zip" --file-in "C:/o3de/test.txt" --file-type="json"
This will put the file at --file-in into the bucket at --full-address if there is not already a file at the desired location

Read:
tiaf_tools.py --s3-bucket "tiaf" --action "read" --full-address "RootDir/Branch/Build/TestSuite/historic_data.json.zip --file-out "C:/o3de/" --file-type="json"
This will download the file at the desired location and store it in the directory specified in --file-out

Update:
tiaf_tools.py --s3-bucket "tiaf" --action "update" --full-address "RootDir/Branch/Build/TestSuite/historic_data.json.zip --file-in "C:/o3de/test.txt" --file-type="json"
This will replace the file at the desired location with the file at --file-in. If no file is in bucket, do nothing.

Delete:
tiaf_tools.py --s3-bucket "tiaf" --action "delete" --full-address "RootDir/Branch/Build/TestSuite/historic_data.json.zip" --file-type="json"
This will delete the file at the specified location if it exists

### CODE REVIEW

**Comment by jonawals** [comment]
Can we get an example usage of the new functionality in the PR body?

File: scripts/build/TestImpactAnalysis/storage_query_tool/local_storage_query_tool.py [hunk 1]
```diff
@@ -106,16 +105,26 @@ def _put(self, file: str, storage_location: str):
...
+                json_obj = json.loads(file)
+                if not self._check_object_exists(str(storage_location)):
+                    with open(f"{storage_location}", "w", encoding="UTF-8") as raw_output_file:
+                        json.dump(json_obj, raw_output_file,
+                                ensure_ascii=False, indent=4)
+                else:
+                    logger.info("Cancelling create, as file exists already")
+            except json.JSONDecodeError:
+                logger.error("The historic data does not contain valid json")
+        if self._file_type == self.FileType.ZIP:
```
Comment (jonawals) [inline scripts/build/TestImpactAnalysis/storage_query_tool/local_storage_query_tool.py:119]: Q: should this not be an else if? I presume the possible values of `_file_type` are mutually exclusive.

File: scripts/build/TestImpactAnalysis/storage_query_tool/local_storage_query_tool.py [hunk 2]
```diff
@@ -106,16 +105,26 @@ def _put(self, file: str, storage_location: str):
...
+            try:
+                if not self._check_object_exists(str(storage_location)):
```
Comment (jonawals) [inline scripts/build/TestImpactAnalysis/storage_query_tool/local_storage_query_tool.py:121]: This logic is virtually identical to the logic of the JSON file type albeit the inner block of the `open`. Can these common code pathways be refactored into one?

File: scripts/build/TestImpactAnalysis/storage_query_tool/local_storage_query_tool.py [hunk 3]
```diff
@@ -124,13 +133,23 @@ def _update(self, file: str, storage_location: str):
...
-            json_obj = json.loads(file)
-            if self._check_object_exists(storage_location):
-                with open(f"{storage_location}", "w", encoding="UTF-8") as raw_output_file:
-                    json.dump(json_obj, raw_output_file,
-                              ensure_ascii=False, indent=4)
-            else:
-                logger.info("Cancelling update, as file does not exist")
-        except json.JSONDecodeError:
-            logger.error("The historic data does not contain valid json")
+        if self._file_type == self.FileType.JSON:
```
Comment (jonawals) [inline scripts/build/TestImpactAnalysis/storage_query_tool/local_storage_query_tool.py:136]: Same here (see prev comment).

File: scripts/build/TestImpactAnalysis/storage_query_tool/s3_storage_query_tool.py [hunk 4]
```diff
@@ -125,19 +125,44 @@ def _access(self, bucket_name: str, file: str, destination: str):
...
-                    file_stream.read()).decode('UTF-8')
-                logger.info(f"Decoding complete.")
-                json_obj = json.loads(decoded_data)
-                with open(f"{destination}historic_data.json", "w", encoding="UTF-8") as raw_output_file:
-                    json.dump(json_obj, raw_output_file,
-                              ensure_ascii=False, indent=4)
+                raise e
+            if self._file_type == self.FileType.JSON:
+                self._save_as_json_file(file_stream, destination)
+            if self._file_type == self.FileType.ZIP:
```
Comment (jonawals) [inline scripts/build/TestImpactAnalysis/storage_query_tool/s3_storage_query_tool.py:131]: Are these also not mutually exclusive conditions?

File: scripts/build/TestImpactAnalysis/tiaf_tools.py [hunk 5]
```diff
@@ -69,10 +68,17 @@ def valid_path(value):
...
     )
 
+    parser.add_argument(
+        "--file-type",
+        choices=["json","zip"],
+        help="What file type SQT should expect to be interacting with. Current options are zip and json.",
+        required=True
+    )
+
     return parser.parse_args()
```
Comment (jonawals) [inline scripts/build/TestImpactAnalysis/tiaf_tools.py:83]: Nit: remove extra line break.

File: scripts/build/TestImpactAnalysis/storage_query_tool/local_storage_query_tool.py (hunk 3 above)
Comment (jcurtisk) [reply scripts/build/TestImpactAnalysis/storage_query_tool/local_storage_query_tool.py:136]: Extracted out some of the code but left the control structure in this method.

**Review by hultonha — APPROVED** [review]
Change looks good to me! Excellent test coverage too, nice work! 😄

**Review by jonawals — APPROVED** [review]

**Pull request merged**